src/libcore/iter/range.rs

```diff
@@ -5,169 +5,288 @@ use crate::usize;
 
-/// Objects that can be stepped over in both directions.
+/// Objects that have a notion of *successor* and *predecessor*
+/// for the purpose of range iterators.
 ///
-/// The `steps_between` function provides a way to efficiently compare
-/// two `Step` objects.
+/// This trait is `unsafe` because implementations of the `unsafe`
+/// trait `TrustedLen` depend on its implementation being correct.
 #[unstable(feature = "step_trait",
-           reason = "likely to be replaced by finer-grained traits",
+           reason = "recently redesigned",
            issue = "42168")]
-pub trait Step: Clone + PartialOrd + Sized {
-    /// Returns the number of steps between two step objects. The count is
-    /// inclusive of `start` and exclusive of `end`.
+pub unsafe trait Step: Clone + PartialOrd + Sized {
+    /// Returns the number of *successor* steps needed to get from `start` to `end`.
     ///
-    /// Returns `None` if it is not possible to calculate `steps_between`
-    /// without overflow.
+    /// Returns `None` if that number would overflow `usize`
+    /// (or is infinite, or if `end` would never be reached).
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`, `b`, and `n`:
+    ///
+    /// * `steps_between(&a, &b) == Some(0)` if and only if `a >= b`
+    /// * `steps_between(&a, &b) == Some(n)` if and only if `a.forward(n) == Some(b)`
+    /// * `steps_between(&a, &b) == Some(n)` if and only if `b.backward(n) == Some(a)`
     fn steps_between(start: &Self, end: &Self) -> Option<usize>;
 
-    /// Replaces this step with `1`, returning itself.
-    fn replace_one(&mut self) -> Self;
-
-    /// Replaces this step with `0`, returning itself.
-    fn replace_zero(&mut self) -> Self;
-
-    /// Adds one to this step, returning the result.
-    fn add_one(&self) -> Self;
+    /// Returns the value that would be obtained by taking the *successor*
+    /// of `self` `count` times.
+    ///
+    /// Returns `None` if this would overflow the range of values supported by `Self`.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`, `n`, and `m` where `n + m` does not overflow:
+    ///
+    /// * `a.forward(n).and_then(|x| x.forward(m)) == a.forward(n + m)`
+    /// * `a.forward(n)` equals `Step::successor` applied to `a` `n` times
+    fn forward(&self, count: usize) -> Option<Self>;
 
-    /// Subtracts one to this step, returning the result.
-    fn sub_one(&self) -> Self;
+    /// Returns the *successor* of `self`.
+    ///
+    /// If this would overflow the range of values supported by `Self`,
+    /// this method is allowed to panic or overflow. Suggested behavior is
+    /// to panic when debug assertions are enabled, and overflow otherwise.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a` where `a.successor()` does not overflow:
+    ///
+    /// * `a == a.successor().unwrap().predecessor().unwrap()`
+    #[inline]
+    fn successor(&self) -> Self { self.forward(1).expect("overflow in `Step::successor`") }
```

Comment:
An additional possible optimization is to make `successor`/`predecessor` `unsafe`, and disallow calling them when it would overflow/underflow. This would allow an implementation on a type with restricted validity (e.g. `char`) to avoid checking the outside bounds at all.

Reply:
I'd love to see some exploration of this, as getting `Range` to use `nuw`/`nsw` now that we have [`unchecked_add`](https://doc.rust-lang.org/beta/std/intrinsics/fn.unchecked_add.html) and friends might help with some of the problematic optimization cases.

My intuition, though, is that it should be a separate
```
unsafe fn successor_unchecked(&self) -> Self { self.successor() }
```
so that safe code can consume the trait, even if it can't implement it.

Reply:
Safe code can always use checked advancing (`forward(1)`).

I'll make them unchecked and suggest safe code to use the checked stepped version. That gives a real reason to having the shortcut ones for single steps.

Reply:
Actually: what does `RangeFrom` do in this world?

`for i in 0u8.. { dbg!(i) }` currently panics in debug mode and wraps infinitely in release mode. If we just make `successor` use `unchecked_add`, then it's UB in both. So I guess `RangeFrom` should consider that it can't go beyond the end of it's value's domain and use `self.start.forward(1).expect("overflow in `Range::next`")` rather than `self.start.successor()`?

Reply:
While safe code _could_ always use `forward(1)`, and that'd be fine in the standard library, if the point of the trait is that it could get stabilized one day, I think having `.successor()` be a thing that works in safe code is important.  And I think your point about `RangeFrom` is a great reason to have distinct `successor` and `successor_unchecked`, since the library will use both.

Reply:
Actually, `RangeFrom` is probably better off using `.forward(1).expect()` anyway, as the current behavior (panic in debug, wrap in release) is debatably a bug and it should always panic. Especially since the panic means that even currently `.nth(n)` is not equivalent to n `.next()`s, as `nth` will panic if the returned value would wrap around. [[playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=335841bad0fca655701eb34ac3c7cbd1)]

Reply:
The real issue with `RangeFrom` is that is panics before returning the largest value (#25708). What it actually needs is either a breaking change, a `wrapping_successor` or a `saturating_successor`.

Reply:
Note that `saturating_successor` can be written as `n.forward(1).unwrap_or(n)`. `successor_unchecked` is "basically" `n.forward(1).unwrap_or_else(unreachable_unchecked)` but easier on LLVM.